test/distributed/shard_tensor/ops/utils.py

```diff
@@ -0,0 +1,139 @@
+# limitations under the License.
+
+from collections.abc import Iterable
+
+import torch
+from torch.distributed.tensor import DTensor, distribute_module
+from torch.distributed.tensor.device_mesh import DeviceMesh
+
+from physicsnemo.distributed import ShardTensor
+
+
+def unparallelize_module(module):
+    """
+    This is the inverse of "distribute_module".  Don't use this except in tests.
+
+    (Why need this?  We're leveraging distribute_module to make sure all
+    ranks have the same weights, if needed, instead of relying on random seeds.)
+    """
+    for name, param in list(module._parameters.items()):
+        if isinstance(param, torch.nn.Parameter) and isinstance(param.data, DTensor):
+            # gather to replicated then unwrap
+            local_tensor = param.data.full_tensor()
+            # replace with a normal Parameter
+            module._parameters[name] = torch.nn.Parameter(
+                local_tensor, requires_grad=param.requires_grad
+            )
+    # recurse into submodules
+    for child in module.children():
+        unparallelize_module(child)
+
+    return module
+
+
+def generate_image_like_data(
+    batch_size: int, C_in: int, spatial_shape: tuple[int, ...]
+) -> torch.Tensor:
+    """
+    Generate a random image-like tensor
+    """
+    return torch.randn(batch_size, C_in, *spatial_shape)
+
+
+def sharded_to_local(container):
+    if isinstance(container, ShardTensor):
+        local_output = container.full_tensor()
+        if container.requires_grad:
+            local_output = local_output.detach().requires_grad_(True)
+        return local_output
+    if isinstance(container, dict):
+        return {key: sharded_to_local(value) for key, value in container.items()}
+    elif isinstance(container, Iterable):
+        return [sharded_to_local(item) for item in container]
+    else:
+        return container
+
+
+def default_tensor_comparison(output, d_output, atol, rtol):
+    # We assume a single output!
+
+    local_output = sharded_to_local(d_output)
+
+    # diff = output - local_output
+    # abs_diff = torch.abs(diff)
+    # max_abs_diff_allowed = 1e-8 + 1e-5 * torch.abs(output)
+
+    # Check forward agreement:
+    assert torch.allclose(output, local_output, atol=atol, rtol=rtol)
+
+    return True
+
+
+def default_loss_fn(output):
+    return output.mean()
+
+
+def numerical_shard_tensor_check(
+    mesh: DeviceMesh,
+    module: torch.nn.Module,
+    input_args: Iterable,
+    input_kwargs: dict,
+    check_grads: bool = False,
+    fwd_comparison_fn: callable = default_tensor_comparison,
+    loss_fn: callable = default_loss_fn,
+    atol: float = 1e-5,
+    rtol: float = 1e-5,
+):
+    # Make sure the module's parameters all align on ever rank of the mesh:
+    d_module = distribute_module(module, device_mesh=mesh)
+    # (By default this replicates)
+
+    # Then, get a local copy of the parameters
+    module = unparallelize_module(module)
```

Comment:
If this is called on vanilla `nn.LayerNorm` immediately after instantiation, is this a no-op? Or should the argument to this function be `d_module`?

Reply:
Good question.  It looks like a typo here but I think it might not be.  The problem is the modules are stateful, and distribute_module returns the module but it is really an _in place_ operation: so `d_module` and `module` point to the same parameter buffers.  I don't love this but it's not my function :/

I'll check this more closely, probably a deepcopy type operation is necessary.